Implement template literal revision

not-an-aardvark opened the pull request:
The [template literal revision proposal](https://github.com/tc39/proposal-template-literal-revision) reached stage 4 in the March 2017 TC39 meeting. This commit adds support for it when `ecmaVersion` is at least 9.

The template literal revision allows invalid escape sequences in tagged template literals, to make it easier to write embedded languages and DSLs. In [ESTree](https://github.com/estree/estree/blob/25834f7247d44d3156030f8e8a2d07644d771fdb/experimental/invalid-tagged-template-escapes.md), this is represented as a `TemplateElement` node with a null `cooked` value.

To allow bad unicode escapes to be tokenized, this updates the template tokenizer to fall back to tokenizing an `invalidTemplate` token if it encounters an invalid escape, rather than immediately throwing an error.

not-an-aardvark commented:
Ping; is there anything I can do to make this easier to review?

(Also, I suppose I should verify that the Acorn team is interested in adding this feature. I'd assumed so because it's a new syntax which has reached stage 4, but let me know if that assumption was incorrect.)

marijnh commented:
Sorry for letting this sit around for so long. I do think it'd be a good addition, but I'm not quite happy with your implementation, especially the use of exceptions to signal a given type of token. Would it be possible to have the tokenizer set the token value to an error marker when it encounters an invalid template, and to have the parser check for this, discarding it when ES>=9 and tagged, and raising it otherwise?

Also, since plugins can access all parser methods, minimizing the amount of breaking changes in method signatures is helpful for plugin writers.

not-an-aardvark commented:
Thanks for the feedback. I'll work on addressing your comments.

> Would it be possible to have the tokenizer set the token value to an error marker when it encounters an invalid template, and to have the parser check for this, discarding it when ES>=9 and tagged, and raising it otherwise?

What kind of error marker do you have in mind? The problem is that `template` tokens store cooked values, so every string is a valid value of a `template` token. As a result, it seems like adding a new token type to store raw values (`invalidTemplate`) is the only clear way to communicate that a template token is invalid. 

This is implemented by having two separate tokenizer methods -- one for producing a cooked value, and one for producing a raw value. If the cooked-value tokenizer encounters an invalid escape sequence, it aborts and the raw-value tokenizer is used instead.

I used exceptions for this because invalid escape sequences are detected by helper methods such as `readEscapedChar` that produce cooked strings for an accumulator. Upon finding an invalid escape sequence, the methods need to communicate that the accumulator should stop collecting cooked strings and delegate to the uncooked-value tokenizer instead. Without changing the return values of `readEscapedChar` and its callers to a result type, there isn't any string return value that could communicate this error.

I'm definitely willing to improve the tokenizer, but I'm unsure how to do it here given that invalid tokens can't have cooked values.

xiemaisi commented on src/tokenize.js:
```diff
@@ -424,7 +424,7 @@ pp.readRegexp = function() {
         tmp = tmp.replace(/\\u\{([0-9a-fA-F]+)\}/g, (_match, code, offset) => {
           code = Number("0x" + code)
-          if (code > 0x10FFFF) this.raise(start + offset + 3, "Code point out of bounds")
+          if (code > 0x10FFFF) this.invalidStringToken(start + offset + 3, "Code point out of bounds")
```
Can `this.inTemplateElement` ever be `true` here?

not-an-aardvark replied:
No, I suppose not. I'll change this back to `raise`.

xiemaisi commented on src/tokenize.js:
```diff
@@ -548,13 +548,38 @@ pp.readString = function(quote) {
+    }
+  }
+
+  this.inTemplateElement = false
```
This will be skipped over if we `throw err` on line 561 above; maybe this should go into a `finally` block?

not-an-aardvark replied:
It's true that this will be skipped over, but does this matter since a parsing error gets thrown anyway? My understanding was that the `Parser` instance is discarded when a syntax error occurs.

xiemaisi replied:
Fair enough.

xiemaisi commented on src/tokenize.js:
```diff
@@ -548,13 +548,38 @@ pp.readString = function(quote) {
 
+const INVALID_TEMPLATE_ESCAPE_ERROR = {}
+
+pp.tryReadTemplateToken = function() {
```
Should this method return something? It currently calls `readTmlToken` and (sometimes) `readInvalidTemplateToken`, but discards the results from both calls.

not-an-aardvark replied:
I'm not very familiar with the tokenizer logic -- is the return value used for anything?

xiemaisi replied:
It looks like `nextToken` may return it: https://github.com/ternjs/acorn/blob/master/src/tokenize.js#L81

not-an-aardvark replied:
It looks like [finishToken](https://github.com/ternjs/acorn/blob/b1cc4ee21af5d36718d651435eb2a2ca94445df4/src/tokenize.js#L179-L187) doesn't return anything though, so I think `readTmplToken` would return `undefined` anyway.

(I'm not trying to argue that there should be no return statement -- I just want to make sure I'm understanding what effects it would have, if any.)

xiemaisi replied:
OK, that makes sense.

marijnh commented:
I've gone ahead and merged this as f1a88fe, since the scope of the exception-control-flow is relatively small, the code looks good overall, and I feel a bit guilty about letting this sit around for so long.
